apps/heft/src/cli/actions/CustomAction.ts

```diff
@@ -0,0 +1,126 @@
+// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
+// See LICENSE in the project root for license information.
+
+import {
+  CommandLineFlagParameter,
+  CommandLineStringParameter,
+  CommandLineIntegerParameter,
+  CommandLineStringListParameter
+} from '@rushstack/ts-command-line';
+
+import { HeftActionBase, IHeftActionBaseOptions } from './HeftActionBase';
+
+/** @alpha */
+export interface ICustomActionParameter {
+  kind: 'flag' | 'integer' | 'string' | 'stringList'; // TODO: Add "choice"
+
+  paramterLongName: string;
+  description: string;
+  callbackValueName: string;
+}
+
+/** @alpha */
+export type CustomActionParameterType = string | boolean | number | ReadonlyArray<string> | undefined;
+
+/** @alpha */
+export interface ICustomActionOptions {
+  actionName: string;
+  documentation: string;
+  summary?: string;
+
+  parameters?: ICustomActionParameter[];
+
+  callback: (parameters: { [callbackValueName: string]: CustomActionParameterType }) => void | Promise<void>;
+}
+
+export class CustomAction extends HeftActionBase {
+  private _customActionOptions: ICustomActionOptions;
+  private _parameterValues: Map<string, () => CustomActionParameterType>;
+
+  public constructor(customActionOptions: ICustomActionOptions, options: IHeftActionBaseOptions) {
+    super(
+      {
+        actionName: customActionOptions.actionName,
+        documentation: customActionOptions.documentation,
+        summary: customActionOptions.summary || ''
+      },
+      options
+    );
+
+    this._customActionOptions = customActionOptions;
+  }
+
+  public onDefineParameters(): void {
+    super.onDefineParameters();
+
+    this._parameterValues = new Map<string, () => CustomActionParameterType>();
+    for (const parameterOption of this._customActionOptions.parameters || []) {
+      if (this._parameterValues.has(parameterOption.callbackValueName)) {
+        throw new Error(`Duplicate callbackValueName: ${parameterOption.callbackValueName}`);
+      }
+
+      let getParameterValue: () => CustomActionParameterType;
+
+      switch (parameterOption.kind) {
+        case 'flag': {
+          const parameter: CommandLineFlagParameter = this.defineFlagParameter({
```

Comment:
Here's a couple other scenarios to think about:

- My plugin implements localization as a preprocessor that runs before the TypeScript compiler task.  Can my plugin add a `--locale LOCALE` parameter to every existing Heft action that invokes the compiler?
- My plugin adds a new action `heft deploy` that performs build+test as part of deploying. Can it reuse some existing parameter definitions (e.g. `--production`, `--lite`, `--max-write-parallelism`) without having to copy+paste the code that defines those parameters?

Somehow this design problem feels very similar to the config file problem... like they should have similar solutions.

But I'm fine with getting a simple solution out while we think through the best design.

Reply:
I think that's a good north star, and I think that'll involve overhauling the whole action system in Heft to do right.

I'd like to get this implementation out and start playing with it, and once we've gained some knowledge and experience with how this all should work, we revisit things.

BTW - I marked these APIs as `@alpha`. Do you know if api-extractor trims `@alpha`s?